### PR TITLE
Kotlin generate union as sealed class

### PR DESCRIPTION
**Pull request by Zomzog**
This PR is for the issue #184

### CODE REVIEW

**Comment by benjamin-bader** [comment]
Hi, thanks for the PR - I'm excited to review it!  Honestly, this is quite a pleasant surprise.  It's a very busy time for me right now, so thanks for your patience.

File: thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt [hunk 1]
```diff
@@ -501,6 +501,53 @@ class KotlinCodeGenerator(
                 .build()
     }
 
+
+    internal fun generateSealedClass(schema: Schema, struct: StructType): TypeSpec {
+        val structClassName = ClassName(struct.kotlinNamespace, struct.name)
+        val typeBuilder = TypeSpec.classBuilder(structClassName).apply {
+            addGeneratedAnnotation()
+
+            if (struct.fields.isNotEmpty()) {
```
[the remark this answers is not in the document]
Comment (Zomzog) [reply thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt]: Changed

File: thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt [hunk 2]
```diff
@@ -501,6 +501,53 @@ class KotlinCodeGenerator(
...
+                addModifiers(KModifier.SEALED)
+            }
+
+            if (struct.isDeprecated) addAnnotation(makeDeprecated())
+            if (struct.hasJavadoc) addKdoc("%L", struct.documentation)
+            if (struct.isException) superclass(Exception::class)
```
[the remark this answers is not in the document]
Comment (Zomzog) [reply thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt]: Changed

File: thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt [hunk 3]
```diff
@@ -352,6 +352,114 @@ class KotlinCodeGeneratorTest {
         """.trimMargin())
     }
 
+
+    private val union: String
```
[the remark this answers is not in the document]
Comment (Zomzog) [reply thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt]: I don't like copy-pasting even in tests but ok it's changed

File: thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt [hunk 4]
```diff
@@ -352,6 +352,114 @@ class KotlinCodeGeneratorTest {
...
+        get() {
+            val thrift = """
+                |namespace kt test.coro
+                |
+                |union Union {
+                |  1: i32 Foo;
+                |  2: i64 Bar;
+                |  3: string Baz
```
[the remark this answers is not in the document]
Comment (Zomzog) [reply thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt]: NotFoo added an test with struct type added

File: thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt [hunk 5]
```diff
@@ -352,6 +352,114 @@ class KotlinCodeGeneratorTest {
...
+            |            this.Foo = null
+            |            this.Bar = null
+            |            this.Baz = value
+            |        }
+            |
+            |        fun build(): Union = when {
+            |            Foo != null -> Foo(Foo)
+            |            Bar != null -> Bar(Bar)
+            |            Baz != null -> Baz(Baz)
+            |            else -> throw AssertionError("unpossible")
```
[the remark this answers is not in the document]
Comment (Zomzog) [reply thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt]: Changed

File: thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt [hunk 6]
```diff
@@ -501,6 +501,53 @@ class KotlinCodeGenerator(
...
+
+            if (parcelize) {
+                addAnnotation(makeParcelable())
```
[the remark this answers is not in the document]
Comment (Zomzog) [reply thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt]: I don't have android SDK to test it, but yes it looks strange for sealed classes. I remove it.

[the remark this answers is not in the document]
Comment (benjamin-bader) [reply thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt]: The KEEP establishing @Parcelize's behavior states that it _should_ work for sealed classes; I'll add it back if/when I verify that it's actually true!

**Comment by benjamin-bader** [comment]
I was curious about the implications of an empty union, so I just checked out out this PR's branch and played around with it.  I see a few problems that we'll need to address before landing this.

- No serialization code is generated for these sealed classes, so any code using unions will be broken.
- There seems to be a bug somewhere (in `thrifty-schema`, probably) relating to empty unions; thrift entities following empty unions seem to be missing from generated code

I found these while attempting to write an integration test in `thrifty-integration-tests`; as a preliminary effort, I added the following to `ClientTestThrift.thrift`:

```thrift
union TestUnion {
  1: i32 AnInt;
  2: i64 ALong;
  3: string Text;
  4: Bonk aBonk;
}

union EmptyUnion {};

struct HasEmptyUnion {
  1: EmptyUnion theEmptyUnion;
}
```

Looking at the generated code in `thrifty-integration-tests/build/generated-src/thrifty`, the absence of an `Adapter` is quite clear.  Less obvious, the `HasEmptyUnion` struct is nowhere to be found.  I can't tell yet why.

Sorry I didn't catch this sooner - I assumed that since integration tests were passing, that everything was working as intended.  It seems we don't have any integration tests at all for unions.

File: thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt [hunk 7]
```diff
@@ -501,6 +501,44 @@ class KotlinCodeGenerator(
...
+                    .superclass(structClassName)
+                    .addProperty(propertySpec.build())
+                    .primaryConstructor(propConstructor.build())
+                    .build()
+            typeBuilder.addType(dataProp)
+        }
+
+        typeBuilder.addType(generateUnionBuilder(schema, struct))
+
+        return typeBuilder
```
Comment (benjamin-bader) [inline thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt]: We need to add a generated `Adapter` here, and also implement the `Struct` interface.

**Comment by Zomzog** [comment]
Adapter and Struct added. I hope it's a little bit better now.

**Comment by Zomzog** [comment]
Hi @benjamin-bader 
Do you have some time for reviewing my PR?
Thanks

**Pull request merged**